modules/Kconfig

```diff
@@ -30,3 +34,23 @@ source "modules/Kconfig.tinycbor"
 source "modules/Kconfig.tinycrypt"
 source "modules/Kconfig.vega"
 source "modules/Kconfig.xtensa"
+
+comment "Unavailable modules, please install those via the project manifest."
+
+# List of Zephyr modules text when module is not available, please use the
+# following syntax:
+# ---------------------------------------------------
+# comment "<module_name> module not available."
```

Comment:
Is it possible to do this automatically by cross referencing the contents of the module roots with the available repos?

Reply:
How ?
Kconfig is not a script language so we cannot loop.

And we need the knowledge somewhere of Zephyr known modules valid but not present, and that is what is obtained by using the `depends on !ZEPHYR_<MODULE_NAME_UPPER>_MODULE`

We prefer not to generate Kconfig from CMake, which is the only place we have the knowledge of module roots and Zephyr handled modules.

Reply:
This just seems like a weird place for an explicit list of modules, something that it seems Zephyr has managed to avoid until now.

I assumed you could generate kconfig in python since that's what you do in zephyr_module.py.

Reply:
Except that this would start to move knowledge of build system into `zephyr_module.py`.
`zephyr_module.py` is dumb in the sense that it simply translates the `module.yml` into meaningful settings for CMake and Kconfig, and not knowing about the build system.

With your suggestion we would suddenly require `zephyr_module.py` to take in additional roots that was provided by user on command line, as well as do internal processing of module roots in module.yml files, and then returning those back to CMake.

This also means that it must be ensured that processing order inside Zephyr modules is identical to processing order in CMake code, creating a vulnerability in the build system.

> something that it seems Zephyr has managed to avoid until now.

Because we have never done something like this before.
Until this PR it has been the responsibility of the user to not enable a module not present, what this PR provides is simply to present whether a module is present or not.

He could enable a module

Reply:
as example, TF-M is presented in NCS v1.4.0, even though it is NOT available, as example:
```
$ cd <path-to>/nrf
$ git checkout v1.4.0
$ west update
....
$ west list trusted-firmware-m
FATAL ERROR: unknown project name/path: trusted-firmware-m
  Hint: use "west list" to list all projects.
$  cmake -DBOARD=nrf9160dk_nrf9160ns -GNinja -Bbuild ../zephyr/samples/hello_world/
$  ninja -Cbuild menuconfig
```

and you can see this:
![image](https://user-images.githubusercontent.com/9802583/102767513-fc6f6b00-437f-11eb-8d25-7aac85b98d60.png)

So even though `trusted-firmware-m` is not a valid module in NCS v1.4.0, it is still presented in menuconfig.

Reply:
Ok, i see that the actual path is passed in at some later point via `$(ZEPHYR_*_KCONFIG)`. At what point are those vars passed to kconfig BTW?

I see why you wouldn't want to do it in zephyr_module.py.

If you'll still humor me with a discussion, I see the kconfig comment as largely cosmetic (informative). Could we provide the information in some other way, like Cmake warnings? AFAIU from modules.cmake, the line

    file(GLOB cmake_modules "${CMAKE_CURRENT_LIST_DIR}/*/CMakeLists.txt")

means that placing a directory in a module root will automatically make the build system expect to find a corresponding module. This means that the kconfig comments in this file are ideally an exact mirror of the list of directories found in module roots. Keeping the same list in two different places is an opportunity for inconsistency to arise. Adding a kconfig comment would be a chore that some people would forget, since everything will build fine without it (AFAIU).

I understand that module roots are passed via the modules themselves, so if a module is missing, the module root might also be missing, i.e. you don't know that anything is missing, but is that a big problem? The places the module is needed should anyway `depends on ZEPHYR_*_MODULE` or something. The way it is here, we also need to maintain a downstream patch of this file.

>So even though trusted-firmware-m is not a valid module in NCS v1.4.0, it is still presented in menuconfig.

But that's handled by the new ZEPHYR_*_MODULE config right? It doesn't really depend on the kconfig comment AFAIU.

Reply:
> I see the kconfig comment as largely cosmetic (informative). 

It is purely informatinve, telling user that a Zephyr module exists, but is not available.

>  Could we provide the information in some other way, like Cmake warnings?

I prefer not, cause that would spam users with all modules not part of their downstream custom west.yml when running CMake. 

With the proposal in the PR, this acts as a purely informative line at the end of the list of modules supported by Zephyr, but not available for various reasons.

And it is only presented to users entering the `menuconfig`.

> Keeping the same list in two different places is an opportunity for inconsistency to arise.

Just as today, where: https://github.com/zephyrproject-rtos/zephyr/blob/4931fc0aef3ca4ca1f58aac99c19034c1fa05e39/modules/Kconfig#L28

must be kept in sync with the file.

so instead of using `source "modules/Kconfig.tfm"` I now ask users to do:
```
comment "TFM module not available."
        depends on !ZEPHYR_TFM_MODULE
```
instead (notice how the `source` line has disappeared, cause it is handled by `zephyr_module.py`).

I ask one more line of code, and exactly same folder / filenames must be kept in sync as current solution.

> i.e. you don't know that anything is missing, but is that a big problem? 

That is exactly the point here, to inform the user that something exists, but is not available (missing).
How will you know that it is missing, if you don't know about it.
Next `west` will start supporting groups: https://github.com/zephyrproject-rtos/west/pull/454
which means there could be optional modules available that user can opt-in to using:
`west config manifest.groups an_optional_group`

>  The way it is here, we also need to maintain a downstream patch of this file.

Why would you need that ?

> 

> >    So even though trusted-firmware-m is not a valid module in NCS v1.4.0, it is still presented in menuconfig.
>
> But that's handled by the new ZEPHYR_*_MODULE config right? It doesn't really depend on the kconfig comment AFAIU.

For the inclusion of the module into the build, that is correct.
The sole purpose of the comment is to INFORM the user entering this menu, that there exists a Zephyr module TFM in vanilla Zephyr but that this module is not available in the workspace.

Reply:
btw. you may decide to leave out:
```
comment "TFM module not available."
        depends on !ZEPHYR_TFM_MODULE
```
and all that will happen is that no information is presented to user when TF-M is not in the workspace.
There could be use-cases where you don't want to list `<my_module>` when it is not present.

Reply:
Ok, if it's optional it's not a big deal for me.

Reply:
> Ok, if it's optional it's not a big deal for me.

Code wise it's optional, but in Zephyr context, I think this should be used unless a given module has good reason why this should not be presented to the user.